datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+
+            # If we set the timeout argument due to a not None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+        cmd: list,
+        protocol_class: type[Protocol],
+        *,
+        cwd: Path | None = None,
+        input: int | IO | bytes | Queue[bytes | None] | None = None,
+        timeout: float | None = None,
+        terminate_time: int | None = None,
+        kill_time: int | None = None,
+) -> Any | Generator:
+
+    introduces_timeout = False
+    if timeout is None:
+        introduces_timeout = True
+        timeout = 1.0
+
+    armed = (
+        False
+        if issubclass(protocol_class, GeneratorMixIn)
+        else True
+    )
+
+    # Create the wrapping class. This is done mainly to ensure that the
+    # termination-related functionality is present in the protocol class that
+    # is used, independent of the actual protocol class that the user passes as
+    # argument.
+    # A side effect of this approach is, that the number of protocol class
+    # definitions is reduced, because the user does not need to define
+    # terminate-capable protocols for every protocol he wants to use.
+    kill_protocol_class = create_kill_wrapper(protocol_class)
+
+    runner = ThreadedRunner(
+        cmd=cmd,
+        protocol_class=kill_protocol_class,
+        stdin=DEVNULL if input is None else input,
+        protocol_kwargs=dict(
+            introduced_timeout=introduces_timeout,
+            terminate_time=terminate_time,
+            kill_time=kill_time,
+            armed=armed,
+            done_future=None,
+            encoding=None
+        ),
+        timeout=timeout,
+        exception_on_error=False,
+        cwd=cwd,
+    )
+    result = runner.run()
+    # We distinguish between a non-generator run, i,e. a blocking run and
+    # a generator run.
+    if not issubclass(protocol_class, GeneratorMixIn):
+        yield result
+    else:
+        try:
+            yield result
+        finally:
+            # Arm the protocol, that will enable terminate signaling or kill
+            # signaling, if terminate_time or kill_time are not None.
+            result.runner.protocol.arm()
+            # Exhaust the generator. Because we have set a timeout, this will
+            # lead to invocation of the timeout method of the wrapper, which
+            # will take care of termination or killing. And it will fetch
+            # the result code of the terminated process.
+            # NOTE: if terminate_time and kill_time are both None, this might
+            # loop forever.
+            tuple(result)
```

Comment:
What is the significance of `tuple()` here. It would collect all the remaining results into memory. This could be a lot.

In my previous attempts I used `proc.communicate()`, which seems to aim for the "same" thing, possibly on a lower level.

Reply:
We do not expect too many results here. An alternative would be to do `for _ in result: pass`. We should exhaust the result generator to trigger cleanup and fetching of the process exit code.

Reply:
If that is always the case (not many results), it would be good to have the comment say how this conclusion is made. From my POV it depends very much on the application and could potentially be huge. So I probably do not understand things properly.

Reply:
You are right, it depends on the application. The idea was that, when we reach this point, we have put the process into a state in which it should exit. Depending on the application in question, there might indeed be a large amount of output, e.g. a summary of operations etc.

I will change it so that it does not store output from the subprocess at this point

Reply:
Done